dhall/src/Dhall/Schemas.hs

```diff
@@ -0,0 +1,189 @@
+import qualified Dhall.Map                                 as Map
+import qualified Dhall.Normalize                           as Normalize
+import qualified Dhall.Optics                              as Optics
+import qualified Dhall.Parser                              as Parser
+import qualified Dhall.Pretty
+import qualified Dhall.Syntax                              as Syntax
+import qualified Dhall.TypeCheck                           as TypeCheck
+import qualified Dhall.Util                                as Util
+import qualified System.AtomicWrite.Writer.LazyText        as AtomicWrite
+import qualified System.Console.ANSI                       as ANSI
+import qualified System.IO                                 as IO
+
+-- | Arguments to the @schemas@ subcommand
+data Schemas = Schemas
+    { characterSet :: CharacterSet
+    , censor       :: Censor
+    , input        :: Input
+    , outputMode   :: OutputMode
+    , schemas      :: Text
+    }
+
+-- | Implementation of the @dhall schemas@ subcommand
+schemasCommand :: Schemas -> IO ()
+schemasCommand Schemas{..} = do
+    originalText <- case input of
+        InputFile file -> Text.IO.readFile file
+        StandardInput  -> Text.IO.getContents
+
+    (Header header, expression) <- Util.getExpressionAndHeaderFromStdinText censor originalText
+
+    schemasRecord <- Core.throws (Parser.exprFromText "(schemas)" schemas)
+
+    schemasExpression <- simplifyUsingSchemas schemasRecord expression
+
+    let docStream =
+            Dhall.Pretty.layout
+                (   Pretty.pretty header
+                <>  Dhall.Pretty.prettyCharacterSet characterSet schemasExpression
+                <>  "\n"
+                )
+
+    let schemasText = Pretty.Text.renderStrict docStream
+
+    case outputMode of
+        Write -> do
+            case input of
+                InputFile file -> do
+                    if originalText == schemasText
+                        then return ()
+                        else AtomicWrite.atomicWriteFile
+                                file
+                                (Pretty.Text.renderLazy docStream)
+                StandardInput -> do
+                    supportsANSI <- ANSI.hSupportsANSI IO.stdout
+
+                    Pretty.Terminal.renderIO
+                        IO.stdout
+                        (if supportsANSI
+                            then fmap Dhall.Pretty.annToAnsiStyle docStream
+                            else Pretty.unAnnotateS docStream)
+
+        Check -> do
+            if originalText == schemasText
+                then return ()
+                else do
+                    let command = "schemas"
+
+                    let modified = "rewritten"
+
+                    Exception.throwIO CheckFailed{..}
+
+decodeSchema
+    :: Expr Src Void
+    -> Maybe (Map Text (Expr Src Void), Map Text (Expr Src Void))
+decodeSchema
+    (RecordLit [ ("Type", Record _Type), ("default", RecordLit _default) ]) =
```

Comment:
That `IsList` instance is very convenient for nested pattern matching!  It's also no different than the one for `Data.Map.Map`

Reply:
The one for `Data.Map` seems more predictable, since it always returns the elements in ascending key order. Maybe that's something that we should adopt…

Reply:
My reasoning for preserving the order is that the `IsList` instance should behave like `Show`, meaning that the literal it produces should be valid code for creating the original value